README.md
# flutter_locations_distance

A Flutter package which provides the distance between two locations
* [See the example folder](https://github.com/Eimji/flutter_locations_distance/tree/master/example)

## Installation

* Add this to your package's pubspec.yaml file:
```
dependencies:
  flutter_locations_distance:
```
* You can install packages from the command line:
  with Flutter:
```
$ flutter packages get
```

* Import it now in your Dart code:
```
 import 'package:flutter_locations_distance/flutter_locations_distance.dart';
```


## Usage
* Code
```
    double distance;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      distance = await FlutterLocationsDistance().distanceBetween(48.8438543, -3.5232399, 48.8589507, 2.2770204);
    } on PlatformException {
      distance = -1.0;
    }
```


## Getting Started with Flutter

For help getting started with Flutter, view our online
[documentation](https://flutter.io/).

For help on editing plugin code, view the [documentation](https://flutter.io/developing-packages/#edit-plugin-package).


CHANGELOG.md
## 0.0.1

* Initial release.

## 0.0.2

* Fix errors due to the command "packages pub publish --dry-run" which gives wrong suggestions

## 0.0.3

* Migration to AndroidX


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
//import 'package:flutter_test/flutter_test.dart';

//import 'package:flutter_locations_distance_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is Text && widget.data.startsWith('Running on:'),
        ),
        findsOneWidget);
  });
}


lib/flutter_locations_distance.dart
import 'dart:async';

import 'package:flutter/services.dart';

class FlutterLocationsDistance {
  factory FlutterLocationsDistance() {
    if (_instance == null) {
      final MethodChannel methodChannel =
          const MethodChannel('flutter.eimji.io/flutter_locations_distance');
      _instance = new FlutterLocationsDistance.private(methodChannel);
    }
    return _instance;
  }

  @visibleForTesting
  FlutterLocationsDistance.private(this._methodChannel);

  static FlutterLocationsDistance _instance;

  final MethodChannel _methodChannel;


  /// Returns the distance between the supplied coordinates in meters.
  Future<double> distanceBetween(double startLatitude, double startLongitude,
    double endLatitude, double endLongitude) =>
      _methodChannel.invokeMethod('distanceBetween', <String, double>{
        "startLatitude": startLatitude,
        "startLongitude": startLongitude,
        "endLatitude": endLatitude,
        "endLongitude": endLongitude
      }).then<double>((dynamic result) => result);
}


ios/Classes/SwiftFlutterLocationsDistancePlugin.swift
import Flutter
import UIKit

import Foundation
import CoreLocation

public class SwiftFlutterLocationsDistancePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter.eimji.io/flutter_locations_distance", binaryMessenger: registrar.messenger())
    let instance = SwiftFlutterLocationsDistancePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "distanceBetween" {
      self.receiveDistanceBetweenLocations(arguments: call.arguments, result: result);
    } else {
      result(FlutterMethodNotImplemented)
    }
  }

  private func receiveDistanceBetweenLocations(arguments: Any?, result: FlutterResult) {
    // parse arguments to get coordiantes
    var _startLocation: CLLocation?
    var _endLocation: CLLocation?
    if let location = arguments as! NSDictionary? {
      let startLatitude = location.object(forKey: "startLatitude") as! CLLocationDegrees
      let startLongitude = location.object(forKey: "startLongitude") as! CLLocationDegrees
      let endLatitude = location.object(forKey: "endLatitude") as! CLLocationDegrees
      let endLongitude = location.object(forKey: "endLongitude") as! CLLocationDegrees
      
      _startLocation = CLLocation.init(latitude: startLatitude, longitude: startLongitude)
      _endLocation = CLLocation.init(latitude: endLatitude, longitude: endLongitude)
    } else {
      _startLocation = nil
      _endLocation = nil
    }
    
    guard let startLocation = _startLocation, let endLocation = _endLocation else {
      result(FlutterError.init(
          code: "ERROR_CALCULATE_DISTANCE_INVALID_PARAMS",
          message: "Please supply start and end coordinates.",
          details: nil))
        
      return
    }
    
    let distance = endLocation.distance(from: startLocation)
    
    result(distance)
  }
}


android/src/main/java/io/eimji/flutterlocationsdistance/FlutterLocationsDistancePlugin.java
package io.eimji.flutterlocationsdistance;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import java.util.Map;
import android.location.Location;

class Coordinate {
    public final double latitude;
    public final double longitude;

    public Coordinate(double latitude, double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }
}

/** FlutterLocationsDistancePlugin */
public class FlutterLocationsDistancePlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter.eimji.io/flutter_locations_distance");
    channel.setMethodCallHandler(new FlutterLocationsDistancePlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("distanceBetween")) {
      receiveDistanceBetweenLocations(call.arguments, result);
    } else {
      result.notImplemented();
    }
  }

  private static void receiveDistanceBetweenLocations(Object arguments, Result flutterResult) {
    // parse arguments to get coordiantes
    Coordinate sourceCoordinate;
    Coordinate destinationCoordinate;

    if(arguments == null) {
        throw new IllegalArgumentException("No coordinates supplied to calculate distance between.");
    }

    @SuppressWarnings("unchecked")
    Map<String, Double> coordinates = (Map<String, Double>)arguments;

    if(coordinates == null)
        throw new IllegalArgumentException("No coordinates supplied to calculate distance between.");

    sourceCoordinate = new Coordinate(
            coordinates.get("startLatitude"),
            coordinates.get("startLongitude"));
    destinationCoordinate = new Coordinate(
            coordinates.get("endLatitude"),
            coordinates.get("endLongitude"));    

    if(sourceCoordinate == null || destinationCoordinate == null) {
        flutterResult.error(
                "ERROR_CALCULATE_DISTANCE_INVALID_PARAMS",
                "Please supply start and end coordinates.",
                null);
    }

    float[] results = new float[1];

    try {
        Location.distanceBetween(
                sourceCoordinate.latitude,
                sourceCoordinate.longitude,
                destinationCoordinate.latitude,
                destinationCoordinate.longitude,
                results);

        // According to the Android documentation the distance is
        // always stored in the first position of the result array
        // (see: https://developer.android.com/reference/android/location/Location.html#distanceBetween(double,%20double,%20double,%20double,%20float[]))
        flutterResult.success(results[0]);
    } catch(IllegalArgumentException ex) {
        flutterResult.error(
                "ERROR_CALCULATE_DISTANCE_ILLEGAL_ARGUMENT",
                ex.getLocalizedMessage(),
                null);
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_locations_distance/flutter_locations_distance.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  double _distance = 0.0;

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    double distance;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      distance = await FlutterLocationsDistance().distanceBetween(48.8438543, -3.5232399, 48.8589507, 2.2770204);
    } on PlatformException {
      distance = -1.0;
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _distance = distance;
    });
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: const Text('Flutter locations distance'),
        ),
        body: new Center(
          child: new Text('Distance between Perros-Guirec and Paris in meters:\n $_distance\n'),
        ),
      ),
    );
  }
}
